### ISSUE TITLE
ComputeEngineCredentials fail when signing

### ISSUE THREAD

**Issue by de55** [googleapis/google-auth-library-java#418]
#### Environment details

- OS: Debian
- Java version: 11
- google-auth-library-java version(s): 0.20.0

#### Steps to reproduce

1. Set up GKE Workload Identity
2. Set up a service account
3. Test signurl using service account key
4. Test signurl using default service account

#### Stacktrace

```
com.google.auth.ServiceAccountSigner$SigningException: Failed to sign the provided bytes at com.google.auth.oauth2.IamUtils.sign(IamUtils.java:87) ...

Caused by: java.io.IOException: Error code 403 trying to sign provided bytes: The caller does not have permission at com.google.auth.oauth2.IamUtils.getSignature(IamUtils.java:125)
```

#### Code snippet

```kotlin
        val googleCredentials = if (key == null) GoogleCredentials.getApplicationDefault() else ServiceAccountCredentials.fromStream(key)
        val storage = StorageOptions.newBuilder().setCredentials(googleCredentials).build().service
        val blobInfo = BlobInfo.newBuilder(BlobId.of(bucketName, fileName)).build()
        val url = storage.signUrl(
            blobInfo,
            expiration,
            TimeUnit.MINUTES,
            Storage.SignUrlOption.withV4Signature()
        )
```

#### Any additional information below

I am trying to create a signed url using ComputeEngineCredentials. When I create ServiceAccountCredentials by providing a key.json, the request works. However, when I use the default ComputeEngineCredentials it fails, even though it is using the same service account.

Should this work using the ComputeEngineCredentials?

Thanks!

**Comment by liqwid** [comment]
@de55 did you manage solve this issue eventually?
Would be very useful for us since we are also facing similar issue with getting  `403 trying to sign provided bytes: The caller does not have permission` error on signature attempt.
Instances have platform scopes enabled and service account has `storage.objects.getIamPolicy` permission.

**Comment by sqlboy** [comment]
I'm having same issue.

**Comment by psylinse** [comment]
I'm having this issue as well. Standard google app engine project can't sign URLs for a bucket.

**Comment by psylinse** [comment]
TLDR: Set the `iam.serviceAccounts.signBlob` by setting `Service Account Token Creator` on the service account. 

Actually I think I just stumbled on the solution while debugging the code. It appears that when you use `ComputeEngineCredentials` it signs with `IamUtils` which makes a call to the IAM api. For whatever reason, the permission to do this is not covered under `Editor` or `Owner` so this was really hard to track down. See this this line:

https://github.com/googleapis/google-auth-library-java/blob/master/oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java#L344-L368

In my case, I was using app engine, but it was using the underlying compute engine service account to do the signing, so I had to set it there: i.e. `xxx-compute@developer.gserviceaccount.com`

This came up when I was trying to sign URLs for the buckets on Google App Engine. I can't find documentation for this anywhere. Is this documented anywhere? This took forever to figure out.

**Comment by rfuhrer** [comment]
@psylinse's solution worked for me.  I happened to be trying to sign a url inside a cloud function so the only difference is that I had to update the function projects IAM account instead of a service account (see https://cloud.google.com/functions/docs/securing/function-identity for what that might be in your case)

Some more stacktrace for anyone else googling this issue:
```
Failed to execute com.bst.energy.Function com.google.auth.ServiceAccountSigner$SigningException: Failed to sign the provided bytes at com.google.auth.oauth2.IamUtils.sign(IamUtils.java:87) at com.google.auth.oauth2.ComputeEngineCredentials.sign(ComputeEngineCredentials.java:361) at com.google.cloud.storage.StorageImpl.signUrl(StorageImpl.java:781)
```

**Comment by worace** [comment]
@psylinse thanks a lot for that explanation -- this took me a while to track down as well.

As @rfuhrer pointed out, the exact service account you need to permission depends on what service you're using. In my case I am also using app engine, so I added the role to the `App Engine default service account` which gcp had created on my behalf.

Here is the `gcloud` command I used in case it helps anyone:

```
gcloud projects add-iam-policy-binding <PROJECT ID> \
    --member serviceAccount:<PROJECT ID>@appspot.gserviceaccount.com \
    --role roles/iam.serviceAccountTokenCreator
```

Once you do that you'll also see it reflected in the IAM UI:

<img width="424" alt="Screen Shot 2020-12-30 at 7 58 31 PM" src="https://user-images.githubusercontent.com/1227440/103393870-eed18800-4ad9-11eb-8b50-94d469ac9147.png">
